File: 54-spiral-matrix/spiral-matrix.js
var spiralOrder = function(matrix) {
    let rows = matrix.length
    let columns = matrix[0].length
    let result = []
    let firstRow = 0
    let lastRow = rows - 1
    let firstColumn = 0
    let lastColumn = columns - 1

    while (result.length < (rows * columns)) {
        // travel from left to right
        for (let i = firstColumn; i <= lastColumn; i++) {
            result.push(matrix[firstRow][i])
        }
        firstRow += 1

        // travel from up to down
        for (let i = firstRow; i <= lastRow; i++) {
            result.push(matrix[i][lastColumn])
        }
        lastColumn -= 1

        if (firstRow <= lastRow) {
            // travel from right to left
            for (let i = lastColumn; i >= firstColumn; i--) {
                result.push(matrix[lastRow][i])
            }
            lastRow -= 1

        }

        if (firstColumn <= lastColumn) {
            // travel from down to up  
            for (let i = lastRow; i >= firstRow; i--) {
                result.push(matrix[i][firstColumn])
            }
            firstColumn += 1
        }

    }
    return result
};

File: 54-spiral-matrix/spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        output = []
        first_row = 0
        first_column = 0
        rows = len(matrix)
        columns = len(matrix[0])

        while (first_row < rows and first_column < columns):
            # traverse left to right
            for j in range(first_column, columns):
                output.append(matrix[first_row][j])
            first_row += 1

            # traverse up to down
            for i in range(first_row, rows):
                output.append(matrix[i][columns - 1])
            columns -= 1

            if (first_row < rows):
                # traverse right to left
                for j in range(columns - 1, first_column - 1, -1):
                    output.append(matrix[rows - 1][j])
                rows -= 1

            if (first_column < columns):
                # traverse bottom to top
                for i in range(rows - 1, first_row - 1, -1):
                    output.append(matrix[i][first_column])
                first_column += 1

        return output